product-service/db/productRepository.ts

```diff
@@ -0,0 +1,92 @@
+import { Client } from 'pg';
+import Product from '../entities/product';
+
+const { PG_HOST, PG_PORT, PG_DATABASE, PG_USERNAME, PG_PASSWORD } = process.env;
+
+const dbOptions = {
+  host: PG_HOST,
+  port: +PG_PORT,
+  database: PG_DATABASE,
+  user: PG_USERNAME,
+  password: PG_PASSWORD,
+  ssl: {
+    rejectUnauthorized: false,
+  },
+  connectionTimeoutMillis: 5000,
+};
+
+async function getProducts(): Promise<Product[]> {
+  const client = new Client(dbOptions);
+  await client.connect();
+
+  try {
+    const { rows: products } = await client.query(`
+          select p.id, p.title, p.description, p.author, p.image, p.price, s.count 
+          from products as p inner join stocks as s on p.id = s.product_id;   
+      `);
+
+    return products as Product[];
+  } finally {
+    client.end();
+  }
+}
+
+async function getProductById(id: string): Promise<Product> {
+  const client = new Client(dbOptions);
+  await client.connect();
+
+  try {
+    const { rows: products } = await client.query(
+      `
+        select p.id, p.title, p.description, p.author, p.image, p.price, s.count 
+        from products as p inner join stocks as s on p.id = s.product_id
+        where p.id = $1;   
+    `,
+      [id]
+    );
+
+    if (products.length == 0) {
+      return null;
+    }
+
+    const product = products[0] as Product;
+    return product;
+  } finally {
+    client.end();
+  }
+}
+
+async function addProduct(productData): Promise<Product> {
+  const client = new Client(dbOptions);
+  await client.connect();
+
+  try {
+    var values = [
+      productData.title,
+      productData.description,
+      productData.price,
+      productData.count,
+    ];
+
+    const { rows } = await client.query(
+      `
+    with inserted_product as (
+      insert into products (title, description, price)
+        values ($1, $2, $3)
+        returning id)
+    insert into stocks (product_id, count)
+    select id, $4 from inserted_product
+    returning product_id as id;  
+    `,
+      values
+    );
```

Comment:
It looks like you don't have transaction.

Reply:
+

Reply:
https://teams.microsoft.com/l/message/19:23a48687f2c7476399cbe296af7e961c@thread.tacv2/1604753686915?tenantId=b41b72d0-4e9f-4c26-8a69-f949f367c91d&groupId=3f6cde28-d4d7-4ba8-9f5a-e6274f579056&parentMessageId=1604753686915&teamName=EPAM%20NodeJS%20in%20AWS%20Cloud&channelName=General&createdTime=1604753686915

Reply:
It's EPAM team, I don't have access to it.

Reply:
Oh, sorry, I thought you are from Epam.

In conversation it was mentioned that the order in CTE is not guaranteed. So, as I understand it would be better to create query with “Begin, Commit, Rollback”

Reply:
Ok, I understand, thanks )